### ISSUE TITLE
Explore KL-regularized policy optimization as an alternative to decoupled PUCT/visit targets

### ISSUE THREAD

**Issue by mintiti** [mintiti/alpharat#110]
## Motivation

AlphaRat has repeatedly hit a structural problem at the interface between noisy search values and policy extraction:

- small or poorly identified differences in estimated Q/payoff values can produce large, discontinuous changes in the selected strategy;
- historically, sparse/degenerate payoff matrices caused arbitrary Nash/tie-breaking behavior and large policy bias;
- the current search still uses decoupled PUCT, i.e. each player independently maximizes `Q + exploration`, then the final policy is visit-proportional;
- project notes already flag that constant-sum interchangeability does **not** imply decoupled PUCT converges to equilibrium marginals.

This suggests reframing search as **regularized policy optimization** rather than treating visit counts as the policy definition.

## Core idea

For a local action-value vector `Q` and reference policy `rho`, define the policy explicitly by

```math
\pi_\tau
= \arg\max_\pi \left[\langle \pi,Q\rangle - \tau D_{KL}(\pi\|\rho)\right]
```

which has the closed form

```math
\pi_\tau(a) \propto \rho(a)\exp(Q(a)/\tau).
```

This changes the `Q -> policy` map from a hard/discontinuous selection rule into a smooth one. A key identity in the simple case is

```math
F_\tau(\pi^*) - F_\tau(\pi)
= \tau D_{KL}(\pi\|\pi^*),
```

so small objective error directly controls policy distance. That is exactly the kind of conditioning AlphaRat has historically lacked when Q estimates are noisy or near-degenerate.

There is related work showing AlphaZero/PUCT can be interpreted approximately as regularized policy optimization, but the proposal here is to make the regularized policy **explicit** and, potentially, make the entire search recursion consistent with that regularized objective.

## Why this may matter specifically for AlphaRat

AlphaRat already exposes/stores the ingredients needed for experiments:

- per-action priors;
- per-action Q values;
- visit counts;
- in MCGS, joint `Q(i,j)` matrices for simultaneous moves.

The current pipeline is roughly

```text
estimate Q
  -> decoupled PUCT argmaxes during search
  -> visit counts
  -> visit-proportional policy
  -> train policy head on that search policy
```

A regularized version could instead become

```text
estimate Q
  -> explicit regularized policy/game solver
  -> stable policy
  -> optionally regularized backup/value target
```

The deepest version would use the joint simultaneous-move matrix directly instead of independently maximizing marginal Qs.

### 1. Root-only diagnostic (minimal risk)

Leave MCTS/MCGS completely unchanged.

At the root, compute

```math
\pi_\tau(a) \propto P(a)\exp(Q_{norm}(a)/\tau)
```

using the existing NN prior `P` and normalized Q values.

Compare against:

- argmax;
- visit-proportional policy;
- current forced-playout-pruned policy.

Diagnostics:

- inject tiny artificial perturbations into Q and measure policy sensitivity;
- measure KL / L1 policy change vs perturbation magnitude;
- benchmark playing strength at fixed search budget;
- sweep `tau` in normalized-value units.

### 2. Change the learning target

Keep ordinary PUCT for search allocation, but train the policy head on the explicit regularized `Q -> policy` solution instead of raw visit counts.

This separates:

- **where search spent computation** from
- **what strategy search currently believes is good**.

Potential benefit: less quantization/noise from low visit counts and smoother targets across near-ties.

### 3. Change search allocation

Use the regularized policy itself to allocate/sample simulations instead of hard `argmax(Q + U)` selection.

This would move AlphaRat closer to explicit regularized policy-optimization search rather than using PUCT as an implicit optimizer.

### 4. Simultaneous-game version (most interesting)

MCGS already stores joint action values `Q(i,j)`. Instead of decoupled marginal PUCT, solve a regularized local game at each node.

For an intentionally zero-sum utility matrix `M`, for example:

```math
(\pi_1^\tau,\pi_2^\tau)
= \arg\max_{\pi_1}\min_{\pi_2}
\left[
\pi_1^T M \pi_2
- \tau D_{KL}(\pi_1\|\rho_1)
+ \tau D_{KL}(\pi_2\|\rho_2)
\right].
```

This would give each node a unique, smooth regularized strategic solution instead of independently maximizing noisy marginals.

For the existing bimatrix/general-sum formulation, the analogous object is closer to an entropy/KL-regularized bimatrix equilibrium / logit QRE, so theoretical guarantees differ and need to be treated separately.

### 5. Consistent regularized backups

The strongest version changes not only action extraction but also the value recursion:

```text
regularized node policy
  -> regularized node value
  -> regularized backup
  -> regularized parent objective
```

Then the value network should learn the regularized continuation value, not the ordinary one.

This is the version that could potentially inherit the strongest `small value error -> small strategy error` guarantees.

## Important distinction: regularization vs coverage

KL regularization improves conditioning, but it does **not** replace exploration/coverage.

If an action is never explored, its Q can still be wrong. Forced playouts, priors, Dirichlet noise, or some explicit coverage mechanism may still be needed.

A useful framing is:

```text
regularization -> stable Q-to-policy conversion
coverage       -> enough information to make Q meaningful
```

## Practical notes

- `tau` should probably operate on `Q / value_scale`, not raw cheese counts, so its meaning is stable across game phase.
- reference policy candidates:
  - NN prior (most natural for AlphaZero-style use);
  - uniform prior for diagnostics;
  - potentially a fixed mixture with uniform to guarantee support.
- the direction of KL matters: forward `KL(pi || rho)` gives the Gibbs/softmax form above; AlphaZero's implicit PUCT regularization is often analyzed with a different KL orientation.
- changing only the final root policy is **not** equivalent to a consistently regularized search/value recursion.

## Success criteria

A first result would already be interesting if explicit regularized extraction:

1. reduces policy sensitivity to tiny Q perturbations;
2. improves low-budget behavior in near-degenerate positions;
3. produces better/stabler policy targets than visits;
4. matches or beats current PUCT at equal NN-eval or wall-clock budgets.

The larger research question is whether AlphaRat should treat MCTS/MCGS as a **tree-structured regularized policy optimizer** rather than a visit-count generator.
